### ISSUE TITLE
Clarification on Evaluation Differences

### ISSUE THREAD

**Issue by kunkunlin1221** [SeungjinJung/GD-FAS#3]
Dear Authors,

I noticed that your evaluation protocol differs considerably from those used in prior FAS works. For example, SSDG (CVPR 2020, https://github.com/taylover-pei/SSDG-CVPR2020/blob/c268920a7408ca78fb425954de2bf5745d1c660a/utils/evaluate.py#L10), which is commonly regarded as a strong baseline, computes the final prediction score by aggregating two frames sampled from the same video before evaluating the metrics.

Could you clarify how your evaluation procedure compares to the SSDG-style protocol, and whether these differences may influence the reported performance?

**Comment by SeungjinJung** [comment]
Dear kunkunlin1221,

Thank you for your interest in our work and for raising this question.
First, we note that there is a slight difference in performance when using different evaluation protocols. 
For a fair comparison with SSDG, we followed the same scoring strategy by aggregating prediction scores from two frames sampled from the same video during evaluation (implementation details are provided below). 
The corresponding results are reported in followed Table.

<table>
  <tr>
    <th rowspan="2">Method</th>
    <th colspan="2">OCI→M</th>
    <th colspan="2">OMI→C</th>
    <th colspan="2">OCM→I</th>
    <th colspan="2">ICM→O</th>
    <th colspan="2">Avg</th>
  </tr>
  <tr>
    <th>HTER</th>
    <th>AUC</th>
    <th>HTER</th>
    <th>AUC</th>
    <th>HTER</th>
    <th>AUC</th>
    <th>HTER</th>
    <th>AUC</th>
    <th>HTER</th>
    <th>AUC</th>
  </tr>
  <tr>
    <td>Single Frame</td>
    <td>0.42</td>
    <td>99.88</td>
    <td>0.93</td>
    <td>99.99</td>
    <td>3.33</td>
    <td>98.61</td>
    <td>2.64</td>
    <td>99.34</td>
    <td>1.83</td>
    <td>99.46</td>
  </tr>
<tr>
<td>Two Random Frame</td>
<td>0.42</td>
<td>99.88</td>
    <td>0.93</td>
    <td>99.96</td>
    <td>3.33</td>
    <td>98.67</td>
    <td>2.92</td>
    <td>99.47</td>
    <td>1.90</td>
    <td>99.45</td>
  </tr>
</table>


As expected, the SSDG protocol is more strict than our default protocol, which leads to a slight performance degradation. Nevertheless, our method still achieves state-of-the-art performance under the SSDG-style evaluation.
That said, we believe the SSDG protocol is less practical in real-world scenarios. 
While protocols similar to ours require only a single image from the current time step of a video during inference, the SSDG approach requires selecting and processing two frames from the video sequence. 
This inevitably introduces additional computational cost and preprocessing overhead.

We adopt the SA-FAS(CVPR2023: https://github.com/sunyiyou/SAFAS/tree/master/datasets) protocol as our primary baseline because SA-FAS was the first method to achieve an HTER below 10 across all evaluation settings, with an average HTER of 7.83. 
In SA-FAS, a single random frame is sampled from each video for evaluation. In contrast, another recent work, FLIP(ICCV 2023:  https://github.com/koushiksrivats/FLIP/tree/main/data/MCIO/txt), selects the first frame of each video for inference.

I hope my answer was helpful to you.

For the random two-frame selection, the revised code is provided below.

data/facedataset.py

41-49 lines
image_x1, image_x2 = self.sample_image(video_name)
image_x1 = self.transform(image_x1)
image_x2 = self.transform(image_x2)

sample = {
"image_x1"  : image_x1,
"image_x2"  : image_x2,
"label"     : spoofing_label,
"domain"    : domain_label,
"name"      : video_name,
}

56-59 lines
image_ids = np.random.choice(range(0, frames_total),2)         
image_path1 = os.path.join(image_dir,frames[image_ids[0]])         
image_path2 = os.path.join(image_dir,frames[image_ids[1]])         
return Image.open(image_path1), Image.open(image_path2)

GD-FAS.py

113-117 lines
image_x1 = test_batch_samples['image_x1'].cuda()
image_x2 = test_batch_samples["image_x2"].cuda()
labels  = test_batch_samples['label'].cuda()
logits1, _ = networks(image_x1)
logits2, _ = networks(image_x2)
probs1 = torch.nn.functional.softmax(logits1, dim=1)
probs2 = torch.nn.functional.softmax(logits2, dim=1)
probs = (probs1+probs2)/2
